contracts/contracts/ExampleWarp.sol

```diff
@@ -0,0 +1,42 @@
+//SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+pragma experimental ABIEncoderV2;
+
+import "./IWarpMessenger.sol";
+
+contract ExampleWarp {
+    address constant WARP_ADDRESS = 0x0200000000000000000000000000000000000005;
+    WarpMessenger warp = WarpMessenger(WARP_ADDRESS);
+
+    // sendWarpMessage sends a warp message to the specified destination chain and address pair containing the payload
+    function sendWarpMessage(
+        bytes32 destinationChainID,
+        bytes32 destinationAddress,
+        bytes calldata payload
+    ) external {
+        warp.sendWarpMessage(destinationChainID, destinationAddress, payload);
```

Comment:
maybe I completely got this wrong but how do we plan other contracts to use these functions in case they want to issue a message? We currently don't return any ID we can store here and we also don't let historical search in `validateWarpMessage` with message IDs(or hashes). Does `validateWarpMessage` returns true if it's called for a later block? I.e warp accepted in x but contract called that in x+1.

Reply:
Currently, the messageID is the hash of the unsigned warp message which is emitted as the un-indexed log data from the relevant warp log.

The current implementation relies on off-chain relayers to ingest these logs, parse the message, and request the signatures from VMs, so the contract emitting the message does not need to know the messageID that it's producing. It would certainly be reasonable to return the messageID to the caller though.

`validateWarpMessage` performs signature verification based off of the P-Chain. It's not related to where the message was produced on this chain since it's verifying a message that originated on a different chain.